### PR TITLE
C++: Implement models for poll, accept and select

### PR DESCRIPTION
**Pull request by MathiasVP**
Final part of https://github.com/github/codeql-c-analysis-team/issues/209 for this iteration.

These are slightly more awkward to model since it's most natural to model some of the flow to the file-descriptor argument, but since these are passed by value (because its value isn't actually changed) they don't have an "output" value after the function returns.

### CODE REVIEW

File: cpp/ql/test/library-tests/dataflow/taint-tests/bsd.cpp [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+int select(int, fd_set *, fd_set *, fd_set *, timeval *);
+
+void test_select(timeval timeout) {
+  fd_set readfds;
+
+  readfds.fd_count = 1;
+  readfds.fd_array[0] = source();
+  select(2, &readfds, nullptr, nullptr, &timeout);
+  sink(&readfds); // $ ast MISSING: ir
```
Comment (geoffw0) [inline cpp/ql/test/library-tests/dataflow/taint-tests/bsd.cpp:63]: Please could you talk me through what's going on with taint in the `poll` and `select` examples?

File: cpp/ql/src/semmle/code/cpp/models/implementations/Accept.qll [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+ * The function `accept` and its assorted variants
+ */
+private class Accept extends ArrayFunction, AliasFunction, TaintFunction, SideEffectFunction {
+  Accept() { this.hasGlobalName(["accept", "accept4", "WSAAccept"]) }
+
+  override predicate hasArrayWithVariableSize(int bufParam, int countParam) {
+    bufParam = 1 and countParam = 2
+  }
+
+  override predicate hasArrayInput(int bufParam) { bufParam = [1, 2] }
```
Comment (geoffw0) [inline cpp/ql/src/semmle/code/cpp/models/implementations/Accept.qll:22]: I think parameter 2 (`socklen_t *addrlen`) is a pointer to a value, not a pointer to a buffer / array.

File: cpp/ql/src/semmle/code/cpp/models/implementations/Poll.qll [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+ */
+private class Poll extends ArrayFunction, AliasFunction, TaintFunction, SideEffectFunction {
+  Poll() { this.hasGlobalName(["poll", "ppoll", "WSAPoll"]) }
+
+  override predicate hasArrayWithVariableSize(int bufParam, int countParam) {
+    bufParam = 0 and countParam = 1
+  }
+
+  override predicate hasArrayInput(int bufParam) {
+    getParameter(bufParam).getUnspecifiedType() instanceof PointerType
```
Comment (geoffw0) [inline cpp/ql/src/semmle/code/cpp/models/implementations/Poll.qll:23]: I'm not convinced that `ppoll`s `*tmo_p` and `*sigmask` are buffer / array parameters (here and in `hasSpecificReadSideEffect`, below).

File: cpp/ql/src/semmle/code/cpp/models/implementations/Select.qll [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+import semmle.code.cpp.models.interfaces.Alias
+import semmle.code.cpp.models.interfaces.SideEffect
+
+/**
+ * The function `select` and its assorted variants
+ */
+private class Select extends ArrayFunction, AliasFunction, TaintFunction, SideEffectFunction {
+  Select() { this.hasGlobalName(["select", "pselect"]) }
+
+  override predicate hasArrayWithUnknownSize(int bufParam) { bufParam = [1 .. 3] }
```
Comment (geoffw0) [inline cpp/ql/src/semmle/code/cpp/models/implementations/Select.qll:18]: Do you not think `nfds` is a good enough expression of the array size?

**Review by geoffw0 — COMMENTED** [review]
A few comments / questions.  Sorry if I've misunderstood anything about the sockets library, its a long time since I've used it.

File: cpp/ql/src/semmle/code/cpp/models/implementations/Select.qll (hunk 4 above)
Comment (MathiasVP) [reply cpp/ql/src/semmle/code/cpp/models/implementations/Select.qll:18]: I don't think so, no. According this [this](https://man7.org/linux/man-pages/man2/select.2.html) description:
> This argument should be set to the highest-numbered file descriptor in any of the three sets, plus 1.

So I guess `fds` could point to a single-element array containing a file descriptor with the value 42, and thus you should set `nfds` to 43. I admit, this is a weird way of doing things, but I'm sure they have/had a good reason for this.

File: cpp/ql/test/library-tests/dataflow/taint-tests/bsd.cpp (hunk 1 above)
Comment (MathiasVP) [reply cpp/ql/test/library-tests/dataflow/taint-tests/bsd.cpp:63]: Sure. In the `poll` case, we taint the file descriptor, and `poll` wait for some I/O event, and then fill the `revents` field with the returned event.

In the `select` case, we taint a file descriptor in `readfds`. [It looks like](https://linux.die.net/man/3/fd_set) this is normally done using the `FD_SET` function which we haven't modeled yet. The call to `select` then taints the `readfds` structure, which we can then use to propagate taint to other `FD_` functions.

Thinking about this taint flow again, I'm actually not sure we want taint to propagate taint to the entire object returned by these functions.

File: cpp/ql/src/semmle/code/cpp/models/implementations/Poll.qll (hunk 3 above)
Comment (MathiasVP) [reply cpp/ql/src/semmle/code/cpp/models/implementations/Poll.qll]: Fixed in 576a87231647cf1e1aebcb222b457bb56de4649a.

File: cpp/ql/src/semmle/code/cpp/models/implementations/Accept.qll (hunk 2 above)
Comment (MathiasVP) [reply cpp/ql/src/semmle/code/cpp/models/implementations/Accept.qll]: Fixed in 576a87231647cf1e1aebcb222b457bb56de4649a.

File: cpp/ql/test/library-tests/dataflow/taint-tests/bsd.cpp (hunk 1 above)
Comment (MathiasVP) [reply cpp/ql/test/library-tests/dataflow/taint-tests/bsd.cpp]: 50dd0c98e9786275b4696cf7a3f7dfcb64913853 removes the taint from these functions.

File: cpp/ql/src/semmle/code/cpp/models/implementations/Select.qll (hunk 4 above)
Comment (geoffw0) [reply cpp/ql/src/semmle/code/cpp/models/implementations/Select.qll:17]: I thought `ndfs` was simply the size of the three pointed-to arrays, but you're right - the parameters are in fact pointers to `fd_set`, a struct containing either a fixed bitmap or a resizable array.  The exact purpose of `ndfs` is still unclear to me, but it isn't the array size, and `hasArrayWithUnknownSize` is appropriate.

**Comment by geoffw0** [comment]
LGTM, though I wouldn't mind a second opinion about taint through `poll` and `select`.

**Review by geoffw0 — APPROVED** [review]
Merging.

**Pull request merged**